### PR TITLE
Grow grass remake

### PR DESCRIPTION
**Pull request by AlemSnyder**
Changes the grass gradient function to run semi-recursively. This significantly increases speed, so much so that the original function was removed. Should be about 14 times faster.

It is possible to multi-process by height, but at this point I don't think it's worth it.

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -159,7 +159,7 @@ int StressTest() {
     std::ifstream biome_file = files::open_data_file("biome_data.json");
     biome_file >> biome_data;
 
-    World world(materials_json, biome_data, 12, 12);
+    World world(materials_json, biome_data, 2, 2);
```
Comment (MrAwesomeRocks) [inline src/main.cpp:162]: Make these into `constexpr`s

File: src/terrain/terrain.hpp [hunk 2]
```diff
@@ -589,6 +589,9 @@ class Terrain {
      */
     uint32_t get_voxel(int x, int y, int z) const;
 
+    inline int get_grass_grad_length() const{
+        return grass_grad_length;
```
Comment (MrAwesomeRocks) [inline src/terrain/terrain.hpp:593]: If this is private it should be underscored
If this is const it should be uppercase

File: src/terrain/terrain_helper.hpp [hunk 3]
```diff
@@ -0,0 +1,123 @@
+// TODO comments
+
+#include "tile.hpp"
+#include "terrain.hpp"
+#include <set>
+
+namespace terrain
+{
+namespace helper{
```
Comment (MrAwesomeRocks) [inline src/terrain/terrain_helper.hpp:9]: Indentation...

File: src/terrain/terrain_helper.hpp [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+            if (x == 0 && y == 0) { continue; }
+            // safety function to test if you xyz is in range;
+            if (!ter.in_range(tile->get_x() + x, tile->get_y() + y, tile->get_z())) { continue; }
+            Tile* adjacent_tile =
+                ter.get_tile(tile->get_x() + x, tile->get_y() + y, tile->get_z());
+            if (edge_detector(adjacent_tile)) {
+                is_source = true;
+                break;
+            }
+        }
```
Comment (MrAwesomeRocks) [inline src/terrain/terrain_helper.hpp:33]: Comment what this does

File: src/terrain/terrain_helper.hpp [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+                        & (getter(adjacent_tile)
+                            < ter.get_grass_grad_length() - 1)) {
+                    setter(adjacent_tile, ter.get_grass_grad_length() - 1);
+                }
+            }
+            setter(tile, ter.get_grass_grad_length());
+            // tile->grow_sink should be set to true
+            next_grass_tiles.insert(tile);
+        }
+    }
```
Comment (MrAwesomeRocks) [inline src/terrain/terrain_helper.hpp:52]: And this

File: src/terrain/terrain_helper.hpp [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    std::set<Tile*> next_grass_tiles;
+    for (Tile* tile : in_grass){
+        for (int x = -1; x < 2; x++)
+        for (int y = -1; y < 2; y++) {
+            if (x == 0 && y == 0) { continue; }
+            // safety function to test if you xyz is in range;
+            if (!ter.in_range(tile->get_x() + x, tile->get_y() + y, tile->get_z())) { continue; }
+            Tile* adjacent_tile = ter.get_tile(tile->get_x() + x, tile->get_y() + y, tile->get_z());
+            //! depends on/low
+            if (adjacent_tile->is_grass() & (getter(adjacent_tile) == height)) {
```
Comment (MrAwesomeRocks) [inline src/terrain/terrain_helper.hpp:76]:
```suggestion
            if (adjacent_tile->is_grass() && (getter(adjacent_tile) == height)) {
```

File: src/terrain/terrain_helper.hpp [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        if (is_source){
+            for (int x = -1; x < 2; x++)
+            for (int y = -1; y < 2; y++) {
+                if (x == 0 && y == 0) { continue; }
+                // safety function to test if you xyz is in range;
+                if (!ter.in_range(tile->get_x() + x, tile->get_y() + y, tile->get_z())) { continue; }
+                Tile* adjacent_tile =
+                    ter.get_tile(tile->get_x() + x, tile->get_y() + y, tile->get_z());
+                if (adjacent_tile->is_grass()
+                        & (getter(adjacent_tile)
```
Comment (MrAwesomeRocks) [inline src/terrain/terrain_helper.hpp:43]:
```suggestion
                        && (getter(adjacent_tile)
```

File: src/terrain/terrain_helper.hpp [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+            if (adjacent_tile->is_grass() & (getter(adjacent_tile) == height)) {
+                next_grass_tiles.insert(adjacent_tile);
+                for (int xn = -1; xn < 2; xn++)
+                for (int yn = -1; yn < 2; yn++) {
+                    if (xn == 0 && yn == 0) { continue; }
+                    // safety function to test if you xyz is in range;
+                    if (!ter.in_range(adjacent_tile->get_x() + xn, adjacent_tile->get_y() + yn, adjacent_tile->get_z())) { continue; }
+                    Tile* adjacent_tile_second = ter.get_tile(adjacent_tile->get_x() + xn, adjacent_tile->get_y() + yn, adjacent_tile->get_z());
+                    //! depends on/low
+                    if (adjacent_tile_second->is_grass() & (getter(adjacent_tile_second) < height-1)) {
```
Comment (MrAwesomeRocks) [inline src/terrain/terrain_helper.hpp:85]:
```suggestion
                    if (adjacent_tile_second->is_grass() && (getter(adjacent_tile_second) < height-1)) {
```

File: src/terrain/terrain_helper.hpp [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+            //! depends on/low
+            if (adjacent_tile->is_grass() & (getter(adjacent_tile) == height)) {
+                next_grass_tiles.insert(adjacent_tile);
+                for (int xn = -1; xn < 2; xn++)
+                for (int yn = -1; yn < 2; yn++) {
+                    if (xn == 0 && yn == 0) { continue; }
+                    // safety function to test if you xyz is in range;
+                    if (!ter.in_range(adjacent_tile->get_x() + xn, adjacent_tile->get_y() + yn, adjacent_tile->get_z())) { continue; }
+                    Tile* adjacent_tile_second = ter.get_tile(adjacent_tile->get_x() + xn, adjacent_tile->get_y() + yn, adjacent_tile->get_z());
+                    //! depends on/low
```
Comment (MrAwesomeRocks) [inline src/terrain/terrain_helper.hpp:84]: Huh? Better comments

File: src/terrain/terrain_helper.hpp [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+                    if (adjacent_tile_second->is_grass() & (getter(adjacent_tile_second) < height-1)) {
+                        //! depends on/low
```
Comment (MrAwesomeRocks) [inline src/terrain/terrain_helper.hpp:86]: Huh? Better comments

**Review by MrAwesomeRocks — CHANGES_REQUESTED** [review]
Is it really THAT hard to type the second `&`?

File: src/main.cpp (hunk 1 above)
Comment (MrAwesomeRocks) [reply src/main.cpp:162]: No the 2s

Comment (AlemSnyder) [reply src/main.cpp:162]: it's literally an r value. I don't think this will change anything.

File: src/terrain/terrain.cpp [hunk 11]
```diff
@@ -33,20 +33,21 @@ Terrain::Terrain() : seed(0) {
...
     const std::map<int, const Material>* material, Json::Value biome_data,
-    std::vector<int> grass_grad_data, int grass_mid_
+    std::vector<int> grass_grad_data, unsigned int grass_mid
 ) {
-    if (grass_mid_ < 0) {
-        grass_mid_ = 0;
+    if (grass_mid >= grass_grad_data.size()) {
+        grass_mid_ = grass_grad_data.size() - 1;
     }
+
```
Comment (MrAwesomeRocks) [inline src/terrain/terrain.cpp:41]: Why not throw an error instead?

File: src/terrain/terrain.cpp [hunk 12]
```diff
@@ -439,28 +438,32 @@ Terrain::init_grass() {
...
 void
-Terrain::grow_grass_low(std::set<Tile*> all_grass){
-    helper::grow_grass_recursive<helper::edge_detector_low, helper::getter_low,
-        helper::setter_low>(*this, all_grass);
+Terrain::grow_grass_low(std::set<Tile*> all_grass) {
+    helper::grow_grass_recursive<
+        helper::edge_detector_low, helper::getter_low, helper::setter_low>(
+        *this, all_grass
+    );
 }
```
Comment (MrAwesomeRocks) [inline src/terrain/terrain.cpp:467]: Put these in a header file and mark them `inline`.

Or if they're private make them `static inline` functions not inside a class.

File: src/terrain/terrain_helper.hpp [hunk 13]
```diff
@@ -1,120 +1,152 @@
...
+// run grow_grass recursive(new_set, grass_grad_length)
+/*
+used to grow the first iteration of of grass
+given a set of all grass tiles gets the grass tiles that adjacent to an
+edge.
+Then next to those tiles set the grass height to max_hight-1.
+*/
+template <bool edge_detector(Tile*), int getter(Tile*), void setter(Tile*, int)>
+void
+grow_grass_recursive(Terrain& ter, std::set<Tile*> all_grass) {
```
Comment (MrAwesomeRocks) [inline src/terrain/terrain_helper.hpp:80]: This has a lot of code duplication with the previous function

**Review by MrAwesomeRocks — CHANGES_REQUESTED** [review]

File: src/terrain/terrain_helper.hpp (hunk 13 above)
Comment (AlemSnyder) [reply src/terrain/terrain_helper.hpp:80]: no?
This function takes all grass, while the above function takes only a line of grass.
They cannot operate the same.

Comment (MrAwesomeRocks) [reply src/terrain/terrain_helper.hpp:80]: Here's a diff without comments

image

Comment (AlemSnyder) [reply src/terrain/terrain_helper.hpp:80]: that is literally three lines. the only reason that is long is because of the formatting.

File: src/terrain/terrain.cpp [hunk 14]
```diff
@@ -452,6 +452,21 @@ Terrain::init_grass() {
...
+}
+
+void
+Terrain::grow_grass_low(std::set<Tile*> all_grass) {
+    helper::grow_grass_recursive<
+        helper::edge_detector_low, helper::getter_low, helper::setter_low>(
+        *this, all_grass
+    );
+}
+
```
Comment (MrAwesomeRocks) [inline src/terrain/terrain.cpp:469]: Why did you move these back and un-inline them?

File: src/terrain/terrain_helper.cpp [hunk 15]
```diff
@@ -0,0 +1,23 @@
+#include "terrain_helper.hpp"
+#include "terrain.hpp"
+#include "tile.hpp"
+
+namespace terrain {
+namespace helper {
+
+Tile* Ninos_horrible_function(Terrain& ter, Tile*tile, int x, int y){
```
Comment (MrAwesomeRocks) [inline src/terrain/terrain_helper.cpp:8]: How about `get_adjacent_tile`

And look at how much nicer the code is now that you've removed duplication.

1. You don't have to grok it twice
2. If you assume its correct now you don't even have to bother reading it

Also you need to re-run your formatter.

**Review by MrAwesomeRocks — CHANGES_REQUESTED** [review]
Look almost done

File: src/terrain/terrain.cpp (hunk 14 above)
Comment (AlemSnyder) [reply src/terrain/terrain.cpp:469]: i, um uh, couldn't um, get it compile...

Comment (AlemSnyder) [reply src/terrain/terrain.cpp:469]: I can't import helper in terrain and import terrain in helper

Comment (MrAwesomeRocks) [reply src/terrain/terrain.cpp:469]: - Move to the top of the file
- Make it just a plain static function
- Mark as inline
- PROFIT

File: .vscode/settings.json [hunk 16]
```diff
@@ -1,4 +1,6 @@
 {
+    "clang-format.executable": "C:/Users/haile/AppData/Local/Programs/Python/Python39/Scripts/clang-format",
```
Comment (MrAwesomeRocks) [inline .vscode/settings.json:2]: Whoops

File: src/terrain/path/tile_iterator.cpp [hunk 17]
```diff
@@ -22,94 +22,85 @@
...
+
+namespace terrain {
+
+namespace path {
+
+static uint8_t offsets[26] = {
+    0b00000000, 0b00000001, 0b00000010, 0b00000100, 0b00000101, 0b00000110, 0b00001000,
+    0b00001001, 0b00001010, 0b00010000, 0b00010001, 0b00010010, 0b00010100, 0b00010110,
+    0b00011000, 0b00011001, 0b00011010, 0b00100000, 0b00100001, 0b00100010, 0b00100100,
+    0b00100101, 0b00100110, 0b00101000, 0b00101001, 0b00101010};
```
Comment (MrAwesomeRocks) [inline src/terrain/path/tile_iterator.cpp:39]:
```suggestion
    0b00100101, 0b00100110, 0b00101000, 0b00101001, 0b00101010,};
```
Add a comma and reformat, it will come out nicer and it helps with better diffs

File: src/terrain/path/tile_iterator.cpp [hunk 18]
```diff
@@ -22,94 +22,85 @@
...
-
-std::array<int8_t, 3> get_indexed_offsets(uint8_t index){
+#include <cstdint>
+#include <utility>
+
+namespace terrain {
+
+namespace path {
+
+static uint8_t offsets[26] = {
```
Comment (MrAwesomeRocks) [inline src/terrain/path/tile_iterator.cpp:35]: Either make this number into a `#define`/add a comment or remove it and just leave the curly braces for the compiler to decide:

```cpp
// This
#define NUM_OFFSETS 26
static uint8_t offsets[NUM_OFFSETS] = /* */

// OR
/* We have 26 offsets because... */
static uint8_t offsets[26] = /* */

// OR
static uint8_t offsets[] = /* */
```

File: src/terrain/terrain_helper.hpp [hunk 19]
```diff
@@ -82,20 +83,19 @@ grow_grass_recursive(Terrain& ter, std::set<Tile*> all_grass) {
...
-                // (in some cases: not solid, and others: solid and not grass)
-                if (edge_detector(adjacent_tile)) {
-                    is_source = true;
-                    break;
-                }
+        for (auto it = ter.get_tile_adjacent_iterator(
+                 ter.pos(tile),
+                 DirectionFlags::HORIZONTAL1 | DirectionFlags::HORIZONTAL2
+             );
+             !it.end(); it++) {
```
Comment (MrAwesomeRocks) [inline src/terrain/terrain_helper.hpp:90]: Move the iterator outside:
```suggestion
        auto it = ter.get_tile_adjacent_iterator(
                 ter.pos(tile),
                 DirectionFlags::HORIZONTAL1 | DirectionFlags::HORIZONTAL2
        );
        for (; !it.end(); it++) {
```

File: src/terrain/terrain_helper.hpp [hunk 20]
```diff
@@ -40,21 +40,22 @@ grow_grass_inner(Terrain& ter, std::set<Tile*> in_grass, int height) {
...
-                    // it should be added to the next iteration
-                    next_grass_tiles.insert(adjacent_tile);
-                    // and the tiles with height less should have their height set
-                    setter(adjacent_tile, height);
-                }
+        for (auto it = ter.get_tile_adjacent_iterator(
+                 ter.pos(tile),
+                 DirectionFlags::HORIZONTAL1 | DirectionFlags::HORIZONTAL2
+             );
+             !it.end(); it++) {
```
Comment (MrAwesomeRocks) [inline src/terrain/terrain_helper.hpp:47]: Move the iterator outside:
```suggestion
        auto it = ter.get_tile_adjacent_iterator(
                 ter.pos(tile),
                 DirectionFlags::HORIZONTAL1 | DirectionFlags::HORIZONTAL2
        );
        for (; !it.end(); it++) {
```

File: src/terrain/terrain_helper.hpp [hunk 21]
```diff
@@ -27,7 +27,7 @@ Then next to those tiles set the grass height to hight-1 if this is higher
 than the saved height.
 */
 
-Tile* Ninos_horrible_function(Terrain& ter, Tile*tile, int x, int y);
+Tile* Ninos_horrible_function(Terrain& ter, Tile* tile, int x, int y);
```
Comment (MrAwesomeRocks) [inline src/terrain/terrain_helper.hpp:30]: If you're not using this then delete the prototype

**Review by MrAwesomeRocks — CHANGES_REQUESTED** [review]

File: src/terrain/path/tile_iterator.cpp [hunk 22]
```diff
@@ -32,11 +32,16 @@ namespace terrain {
 
 namespace path {
 
+/* We have 26 offsets because 3^3-1 = 26
+   The total number of adjacent cubes is the number cubes in a three by three
+   area minus one for the central cube.
+ */
```
Comment (MrAwesomeRocks) [inline src/terrain/path/tile_iterator.cpp:38]: Ah so it's a lookup table. Got it
```suggestion
/* We have 26 offsets because 3^3-1 = 26
 * The total number of adjacent cubes is the number cubes in a three by three
 * area minus one for the central cube.
 */
```

File: .vscode/settings.json [hunk 23]
```diff
@@ -1,6 +1,4 @@
 {
-    "clang-format.executable": "C:/Users/haile/AppData/Local/Programs/Python/Python39/Scripts/clang-format",
-    "editor.formatOnSave": true,
```
Comment (MrAwesomeRocks) [inline .vscode/settings.json]: Keep `"editor.formatOnSave": true`

**Review by MrAwesomeRocks — CHANGES_REQUESTED** [review]

File: src/terrain/terrain.cpp (hunk 14 above)
Comment (AlemSnyder) [reply src/terrain/terrain.cpp]: If you can figure out how to do this with out eight forward declarations be my guest.

File: src/main.cpp (hunk 1 above)
Comment (MrAwesomeRocks) [reply src/main.cpp:162]: Good practice for explaining what arbitrary constants mean.

File: src/terrain/terrain.cpp (hunk 14 above)
Comment (MrAwesomeRocks) [reply src/terrain/terrain.cpp]: > terrain.hpp? then I would need to import terrain_helper or make a forward definition of helper::edge_detector

No, `terrain.cpp`. Its only called from within that file.

> ? static? This is a stand alone function. I think this is the same as inline.

No you need both they don't imply each other.

File: src/terrain/terrain_helper-impls.cpp [hunk 24]
```diff
@@ -0,0 +1,47 @@
...
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ */
+
+//! This is an implementations file. Anything not implemented below will lead
+//! linker errors.
```
Comment (MrAwesomeRocks) [inline src/terrain/terrain_helper-impls.cpp:15]:
```suggestion
//! to linker errors.
```

File: src/terrain/path/node-impls.cpp [hunk 25]
```diff
@@ -1,6 +1,10 @@
+
+//! This is an implementations file. Anything not implemented below will lead
+//! linker errors.
```
Comment (MrAwesomeRocks) [inline src/terrain/path/node-impls.cpp:3]:
```suggestion
//! to linker errors.
```

**Review by MrAwesomeRocks — CHANGES_REQUESTED** [review]
Alright this is a better implementation, just 
- fix the grammer
- add the constexprs
- yeet the commented code

**Comment by MrAwesomeRocks** [comment]
Can you just run the formatter once more?

See the readme for directions.

**Review by MrAwesomeRocks — CHANGES_REQUESTED** [review]
Now go through and remove all commented out *code*, like L55 of terrain.hpp

And if code needs to stay commented out, instead of `/* */` use `#if 0`

**Review by MrAwesomeRocks — APPROVED** [review]
LGTM
